### ISSUE TITLE
Unit tests - bugs and suggestions

### ISSUE THREAD

**Issue by ptaconet** [ptaconet/modisfast#15]
Hi @ptaconet, I have some feedback regarding the unit tests. Please let me know if you would prefer I open a separate issue for this, I will gladly do so.

1. Note: It is recommended, though not strictly necessary, that unit tests be structured in the same way as the files under R/; that is, one unit test per file/function. Among other things, this makes it easy to have a quick overview of which major functions are being tested and it also would allow you to check the level of testing coverage with a package like `covr`.
2. Arguments in unit tests should be explicitly named (e.g. `mf_login(credentials = c(earthdata_un, earthdata_pw))` rather than `mf_login(c(earthdata_un,earthdata_pw))`). This helps with readability and clarity but also makes the tests more robust in case new arguments are added to a function or the order of the arguments is changed. Line breaks would also be benefitial for readability.
3. This is more so a suggestion. Though some of the features do require credentials to be tested, there are others that could be tested without the use of credentials. For example, some of the sanity checks in mf_login() can be triggered by providing obviously invalid input: 
```
> mf_login(credentials = "hello")
Error in mf_login(credentials = "hello") : 
  credentials must be a vector character string of length 2 (username and password)
```
Adding these where possible and skipping only the tests that absolutely cannot run might allow for some degree of CI integration. [This testthat article](https://testthat.r-lib.org/articles/skipping.html#helpers) might be helpful. Also, in case it is useful, I see there are some tools available to test functions that use credentials or API requests: https://www.reddit.com/r/rstats/comments/n42yr4/how_to_handle_secret_data_in_unit_tests_of/?rdt=64315 

4. Some tests could be more extensive. For example, the tests for mf_list_collections() only checks that it is a data.frame, but there are no checks for the names and structure of the dataframe.

5. Failed tests after following the testing instructions:
```
── Failed tests ───────────────────────────────────────────────────────────────────────────────────────
Failure (test-modisfast.R:16:3): function mf_login() sends back the expected and outputs and errors when necessary
mf_login(c(earthdata_un, earthdata_pw)) is not NULL

`actual` is a character vector ('cli-8111-557')
`expected` is NULL

Failure (test-modisfast.R:207:3): modisfast works to download a GPM datacube
sum(res_dl$fileSize) not equal to 5689873.
1/1 mismatches
[1] 5564183 - 5689873 == -125690

[ FAIL 2 | WARN 283 | SKIP 0 | PASS 70 ]
```
Also, both the GPM datacube and the MODIS datacube raise several of these warnings on my laptop:
`GDAL Message 1: No UNIDATA NC_GLOBAL:Conventions attribute`. It seems like this might be related to the datasets and not a problem with modisfast/my installation but I have limited experience with GDAL so I am bringing it up just in case.

Cheers,

Victòria

_Originally posted by @V-Agudetse in https://github.com/ptaconet/modisfast/issues/10#issuecomment-2451711704_

**Comment by ptaconet** [comment]
@V-Agudetse opening your comments as a new issue

**Comment by V-Agudetse** [comment]
Thanks @ptaconet! Linking https://github.com/openjournals/joss-reviews/issues/7343#issuecomment-2447717158

**Comment by ptaconet** [comment]
Hi @V-Agudetse ! Thanks again for your suggestions.

I have updated the unit tests following your recommendations. I have tried to follow them all as much as possible, except for n°3. The main functions of the package actually require credentials, and I think I prefer being homogeneous in the way I propose users to run the tests.

I have also updated in the ReadMe the instructions on how to run the tests after installation of the package : see https://github.com/ptaconet/modisfast#manual-testing-of-the-functionality

Regarding the warning GDAL Message 1: No UNIDATA NC_GLOBAL:Conventions attribute , it is in fact more related to your installation of GDAL I believe. 

Do not hesitate to test and tell me if you are ok with this new version, so that I can either propose new modification to the tests or close the issue !
